File: server/go/pkg/tcp_server.go
package pkg

import (
	"bufio"
	"fmt"
	"net"
	"os"
	"time"
)

type TcpServerer interface {
}

type User struct {
	Username   string
	LastLogin  time.Time
	Connection net.Conn
	isOnline   bool
}

func NewUser(username string, Connection net.Conn) *User {
	return &User{
		Username:   username,
		LastLogin:  time.Now(),
		Connection: Connection,
		isOnline:   true,
	}
}

func (u *User) SetOnline(online bool) {
	u.isOnline = online
}

func (u *User) IsOnLine() bool {
	return u.isOnline
}

type TcpServer struct {
	host     string
	port     int
	users    map[string]*User
	listener net.Listener
}

func NewTcpServer(host string, port int) *TcpServer {
	return &TcpServer{
		host:  host,
		port:  port,
		users: make(map[string]*User),
	}
}

func (t *TcpServer) StartInAThread() error {
	go func() {
		err := t.Start()
		if err != nil {
			fmt.Fprintf(os.Stderr, "error starting TCP server: %v\n", err)
		}
	}()
	return nil
}
func (t *TcpServer) Start() error {
	address := fmt.Sprintf("%s:%d", t.host, t.port)
	listener, err := net.Listen("tcp", address)
	if err != nil {
		return fmt.Errorf("error starting TCP server: %v", err)
	}
	t.listener = listener

	fmt.Printf("Server started on %s\n", address)

	for {
		conn, err := listener.Accept()
		if err != nil {
			fmt.Fprintf(os.Stderr, "error accepting connection: %v\n", err)
			break
		}
		go t.handleConnection(conn)
	}

	fmt.Printf("Server stopped\n")
	return nil
}

func (t *TcpServer) Stop() error {
	return t.listener.Close()
}

func (t *TcpServer) handleConnection(conn net.Conn) {
	defer conn.Close()
	reader := bufio.NewReader(conn)
	writer := bufio.NewWriter(conn)
	var user *User
	for {
		header := &ChatHeader{}
		err := header.Read(reader)
		if err != nil {
			fmt.Fprintf(os.Stderr, "error reading header: %v\n", err)
			break
		}

		fmt.Printf("Received header: %+v\n", header)

		switch header.Key() {
		case CommandLoginKey:
			login := &CommandLogin{}
			err := login.Read(reader)
			if err != nil {
				fmt.Fprintf(os.Stderr, "error reading login: %v\n", err)
				break
			}
			fmt.Printf("Received login: %+v\n", login)
			genericResponse := NewGenericResponse(login.CorrelationId(), ResponseCodeOk)
			user = NewUser(login.Username(), conn)
			t.users[login.Username()] = user
			_, err = genericResponse.Write(writer)
			if err != nil {
				fmt.Fprintf(os.Stderr, "error writing response: %v\n", err)
				break
			}
			err = writer.Flush()
			if err != nil {
				fmt.Fprintf(os.Stderr, "error flushing writer: %v\n", err)
				break
			}

		case CommandMessageKey:
			// handle command 2
		}

	}
	if user != nil {
		user.SetOnline(false)
	}

}

func (t *TcpServer) Users() map[string]*User {
	return t.users
}

File: server/rust/src/tcp_server.rs
use std::error::Error;
use bytes::BytesMut;
// use tokio::net::{TcpListener, TcpStream};
use tokio::io::{AsyncReadExt, AsyncWriteExt, BufReader};
use tokio::net::{TcpListener, TcpStream};
use crate::codec::{Decoder, Encoder};
use crate::commands::login::{LoginRequest, LoginResponse};
use crate::response::response::{Response, ResponseCode};
use crate::response::response::ResponseKind::Login;
use crate::types::Header;
use crate::types::version::{COMMAND_LOGIN, GENERIC_RESPONSE, PROTOCOL_VERSION};

#[derive(PartialEq, Eq, Debug)]
pub struct User {
    pub(crate) user_name: String,
}

#[derive(PartialEq, Eq, Debug)]
pub struct Users {
    pub(crate) users: Vec<User>,
}

impl Users {
    pub fn push(&mut self, user: User) {
        self.users.push(user);
    }
}


impl Users {
    pub fn new() -> Users {
        Users {
            users: Vec::new()
        }
    }
}


#[derive(PartialEq, Eq, Debug)]
pub struct TcpServer {
    pub(crate) users: Users,
}

impl TcpServer {
    pub fn new() -> TcpServer {
        TcpServer {
            users: Users::new()
        }
    }

    pub async fn start(&mut self) -> std::io::Result<()> {
        let listener = TcpListener::bind("127.0.0.1:5555").await?;
        println!("Server listening on 127.0.0.1:5555");
        loop {
            let (socket, _) = listener.accept().await?;
            tokio::spawn(async move {
                if let Err(e) = handle_client(socket, &self.users).await {
                    println!("Error handling client: {}", e);
                }
            });
        }
    }
}

async fn handle_client(mut socket: TcpStream,  users: &Users) -> Result<(), Box<dyn Error>> {
    let (reader, mut writer) = socket.split();
    let mut reader = BufReader::new(reader);
    loop {
        let mut buffer = BytesMut::new();
        let mut len_bytes = [0u8; 4];
        reader.read_exact(&mut len_bytes).await?;

        let len = u32::from_be_bytes(len_bytes) as usize;

        // Read the data.  This will wait until the specified number of bytes is available.
        buffer.resize(len, 0);
        reader.read_exact(&mut buffer).await?;

        match Header::decode(&buffer) {
            Err(e) => {
                println!("Error decoding header: {:?}", e);
                break;
            }
            Ok(v) => {
                let (h, hq) = v;
                match hq.key() {
                    COMMAND_LOGIN => {
                        let (_, login) = LoginRequest::decode(h).unwrap();
                        println!("Logged{}", login.user_name);
                        let response = Response::new(
                            Header::new(PROTOCOL_VERSION, GENERIC_RESPONSE),
                            Login(LoginResponse::new(login.correlation_id, ResponseCode::Ok)),
                        );
                        writer.write_all(response_buffer(&response).await.unwrap().as_slice()).await?;
                    }
                    _ => {
                        println!("error");
                        break;
                    }
                }
            }
        };
        writer.flush().await?;
    }
    Ok(())
}

async fn response_buffer(response: &Response) -> Result<Vec<u8>, Box<dyn Error>> {
    let mut writer_tmp = Vec::new();
    writer_tmp.write_u32(response.encoded_size()).await?;
    response.encode(&mut writer_tmp).unwrap();
    writer_tmp.flush().await?;
    Ok(writer_tmp)
}




